pkg/approver/aws/csr.go

```diff
@@ -0,0 +1,184 @@
+		successMessage: "kube-aws-approver approved new node client cert",
+	}}
+}
+
+func isNodeClientCert(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !reflect.DeepEqual([]string{"system:nodes"}, x509cr.Subject.Organization) {
+		return false
+	}
+	if (len(x509cr.DNSNames) > 0) || (len(x509cr.EmailAddresses) > 0) || (len(x509cr.IPAddresses) > 0) {
+		return false
+	}
+	if !hasExactUsages(csr, kubeletClientUsages) {
+		return false
+	}
+	if !strings.HasPrefix(x509cr.Subject.CommonName, "system:node:") {
+		return false
+	}
+	return true
+}
+
+func isSelfNodeClientCert(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !isNodeClientCert(csr, x509cr) {
+		return false
+	}
+	if csr.Spec.Username != x509cr.Subject.CommonName {
+		return false
+	}
+	return true
+}
+
+func (ar *Approver) isValidNewNode(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	idu := getInstanceIDFromUsername(csr.Spec.Username)
+	nn := getNodeNameFromCN(x509cr.Subject.CommonName)
+	idn, err := ar.aws.instanceID(nn)
+	if err != nil {
+		return false
+	}
+	if idn != idu {
+		return false
+	}
+	_, err = ar.kubeClient.CoreV1().Nodes().Get(nn, metav1.GetOptions{})
+	if !apierrors.IsNotFound(err) {
+		return false
+	}
+	return true
+}
+
+func (ar *Approver) isValidNode(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	nn := getNodeNameFromCN(x509cr.Subject.CommonName)
+	_, err := ar.aws.instanceID(nn)
+	if err != nil {
+		return false
+	}
+	return true
+}
+
+func (ar *Approver) isValidASG(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	nn := getNodeNameFromCN(x509cr.Subject.CommonName)
+	asg, err := ar.aws.autoScalingGroupID(nn)
+	if err != nil {
+		return false
+	}
+	return ar.allowedASGs.Has(asg)
+}
+
+func getCertApprovalCondition(status *certificates.CertificateSigningRequestStatus) (approved bool, denied bool) {
+	for _, c := range status.Conditions {
+		if c.Type == certificates.CertificateApproved {
+			approved = true
+		}
+		if c.Type == certificates.CertificateDenied {
+			denied = true
+		}
+	}
+	return
+}
+
+var kubeletClientUsages = []certificates.KeyUsage{
+	certificates.UsageKeyEncipherment,
+	certificates.UsageDigitalSignature,
+	certificates.UsageClientAuth,
+}
+
+func hasExactUsages(csr *certificates.CertificateSigningRequest, usages []certificates.KeyUsage) bool {
+	if len(usages) != len(csr.Spec.Usages) {
+		return false
+	}
+
+	usageMap := map[certificates.KeyUsage]struct{}{}
+	for _, u := range usages {
+		usageMap[u] = struct{}{}
+	}
+
+	for _, u := range csr.Spec.Usages {
+		if _, ok := usageMap[u]; !ok {
+			return false
+		}
+	}
+
+	return true
+}
+
+func getNodeNameFromCN(cn string) string {
+	return strings.TrimPrefix(cn, "system:node:")
+}
+
+func getInstanceIDFromUsername(username string) string {
+	return strings.TrimPrefix(username, "system:boostrappers:")
```

Comment:
need error checking here as well.

Also how is this configured client/server side? Please add some comments.

Reply:
Added some comments. Also all recognizer funcs have comments too.